realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -94,20 +102,78 @@
     // we register the listener.
     private final AtomicLong progressListenerId = new AtomicLong(-1);
 
+    // List of Java state change listeners
+    private final CopyOnWriteArrayList<SessionStateListener> sessionStateListeners = new CopyOnWriteArrayList<>();
+
+    // Reference to the token representing the native state change
+    // Only one listener is used for all Java listeners.
+    private long nativeStateListenerToken;
+
     // represent different states as defined in SyncSession::PublicState 'sync_session.hpp'
     private static final byte STATE_VALUE_WAITING_FOR_ACCESS_TOKEN = 0;
     private static final byte STATE_VALUE_ACTIVE = 1;
     private static final byte STATE_VALUE_DYING = 2;
     private static final byte STATE_VALUE_INACTIVE = 3;
     private static final byte STATE_VALUE_ERROR = 4;
 
+    // List of Java connection change listeners
+    private final CopyOnWriteArrayList<ConnectionListener> connectionListeners = new CopyOnWriteArrayList<>();
+
+    // Reference to the token representing the native listener for connection changes
+    // Only one native listener is used for all Java listeners
+    private long nativeConnectionListenerToken;
+
+    // represent different states as defined in SyncSession::PublicConnectionState 'sync_session.hpp'
+    // saved here instead of as constants in ConnectionState.java to enable static checking by JNI
+    static final byte CONNECTION_VALUE_DISCONNECTED = 0;
+    static final byte CONNECTION_VALUE_CONNECTING = 1;
+    static final byte CONNECTION_VALUE_CONNECTED = 2;
+
     private URI resolvedRealmURI;
 
+    /**
+     * Enum describing the states a SyncSession can be in. The initial state is
+     * {@link State#INACTIVE}.
+     * <p>
+     * A Realm will automatically synchronize data with the server if the session is either {@link State#ACTIVE}
+     * or {@link State#DYING} and {@link #isConnected()} returns {@code true}.
```

Comment:
AFAIK `State#DYING` depends on the sync session closing strategy (currently default to `SyncSessionStopPolicy::AfterChangesUploaded`) this will change if we expose the stop policy strategy

Reply:
Yes, this is true. This is outside the user's control for now. I modified it so it says it always transition to dying.